### PR TITLE
NASA daily image UI #2

### PR DESCRIPTION
**Pull request by SergAndroid**

### CODE REVIEW

File: app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt [hunk 1]
```diff
@@ -17,32 +25,105 @@ class MainActivity : AppCompatActivity() {
...
+        }
+        return super.onOptionsItemSelected(item)
+    }
 
     /**
      * Don't edit this function. Use only for fetching data.
      */
     private fun fetchDailyImage(date: Date? = null) {
+        if (!coroutineScope.isActive) coroutineScope = CoroutineScope(Dispatchers.IO)
+
```
Comment (Assasinnys) [inline app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt:205]: Explain the reason of changing this.

Comment (SergAndroid) [reply app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt:205]: It need to reinit coroutineScope after cancel

Comment (Assasinnys) [reply app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt:205]: Sure

**Comment by Assasinnys** [comment]
If connection is lost, but user uploaded screen, he will see this bug.
connection_bug

File: app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt [hunk 2]
```diff
@@ -17,32 +25,105 @@ class MainActivity : AppCompatActivity() {
...
+                } else {
+                    val newDateInMillis = newDate.timeInMillis
+                    fetchDailyImage(Date(newDateInMillis))
+                }
+
+            },
+            currentYear,
+            currentMonth,
+            currentDay
+        ).show()
```
Comment (Assasinnys) [inline app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt]: It's better to disable future dates. DatePicker can do this.

**Review by Assasinnys — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt (hunk 2 above)
Comment (SergAndroid) [reply app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt]: Ok, done.

File: app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt [hunk 3]
```diff
@@ -40,50 +46,106 @@ class MainActivity : AppCompatActivity() {
...
+                val newDateInMillis = newDate.timeInMillis
+                fetchDailyImage(Date(newDateInMillis))
             },
             currentYear,
             currentMonth,
             currentDay
-        ).show()
+        )
+        datePickerDialog.datePicker.maxDate = System.currentTimeMillis()
+        datePickerDialog.show()
```
Comment (Assasinnys) [inline app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt]: Good option is to use apply{} function

File: app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt [hunk 4]
```diff
@@ -40,50 +46,106 @@ class MainActivity : AppCompatActivity() {
...
+    private fun showLoading(b: Boolean) {
+        with(binding) {
+            if (b) {
+                contentLayout.alphaFade(!b)
+                avLoading.alphaFade(b)
+                avLoading.playAnimation()
+            } else {
+                avLoading.alphaFade(b)
+                avLoading.cancelAnimation()
+                contentLayout.alphaFade(!b)
```
Comment (Assasinnys) [inline app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt]: You could make it shorter using functions such as apply, also, run, let, with

**Review by Assasinnys — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt [hunk 5]
```diff
@@ -87,17 +83,16 @@ class MainActivity : AppCompatActivity() {
     }
 
     private fun showNetworkError(b: Boolean) {
```
Comment (Assasinnys) [inline app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt]: Bad name of the parameter

File: app/src/main/java/com/example/basiccomponents/extensions/ViewExt.kt [hunk 6]
```diff
@@ -9,14 +9,15 @@ import androidx.core.view.isVisible
 fun View.alphaFade(b: Boolean) {
```
Comment (Assasinnys) [inline app/src/main/java/com/example/basiccomponents/extensions/ViewExt.kt]: Bad name of the parameter

File: app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt [hunk 7]
```diff
@@ -40,50 +46,106 @@ class MainActivity : AppCompatActivity() {
...
             currentDay
-        ).show()
+        )
+        datePickerDialog.datePicker.maxDate = System.currentTimeMillis()
+        datePickerDialog.show()
     }
 
     private fun handleError(t: Throwable) {
         Log.e(MainActivity::class.java.simpleName, "exception! CurrentThread: ${Thread.currentThread().name}", t)
         //TODO: handle request's errors here
```
Comment (Assasinnys) [inline app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt]: Remove comments from your code except the doc comments

**Review by Assasinnys — CHANGES_REQUESTED** [review]
Looks good. Small details left.

File: app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt [hunk 8]
```diff
@@ -82,36 +81,36 @@ class MainActivity : AppCompatActivity() {
         }
     }
 
-    private fun showNetworkError(b: Boolean) {
+    private fun showNetworkError(showError: Boolean) {
```
Comment (Assasinnys) [inline app/src/main/java/com/example/basiccomponents/ui/dailyimage/MainActivity.kt]: Ohh. Name of a boolean parameters should start with "is", "are", "has".

**Review by Assasinnys — CHANGES_REQUESTED** [review]

**Review by Assasinnys — CHANGES_REQUESTED** [review]
When daily is uploaded without an internet connection, the satellite overlaps nasa image. Also the satellite stays on the screen.
Screenshot_1626739324

File: app/src/main/java/com/example/basiccomponents/extensions/ViewExt.kt [hunk 9]
```diff
@@ -6,8 +6,8 @@ import androidx.activity.OnBackPressedCallback
 import androidx.core.animation.doOnEnd
 import androidx.core.view.isVisible
 
-fun View.alphaFade(b: Boolean) {
-    val start = if (b) 0F else 1F
+fun View.alphaFade(b: Boolean): ValueAnimator? {
+    val start = this.alpha
     val end = if (b) 1F else 0F
```
Comment (Assasinnys) [inline app/src/main/java/com/example/basiccomponents/extensions/ViewExt.kt]: Again. Boolean name is invalid

**Review by Assasinnys — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/basiccomponents/extensions/ViewExt.kt [hunk 10]
```diff
@@ -2,13 +2,10 @@ package com.example.basiccomponents.extensions
...
 import android.view.View
-import androidx.activity.OnBackPressedCallback
-import androidx.core.animation.doOnEnd
-import androidx.core.view.isVisible
 
-fun View.alphaFade(b: Boolean): ValueAnimator? {
+fun View.alphaFade(show: Boolean): ValueAnimator? {
     val start = this.alpha
-    val end = if (b) 1F else 0F
+    val end = if (show) 1F else 0F
```
Comment (Assasinnys) [inline app/src/main/java/com/example/basiccomponents/extensions/ViewExt.kt]: It's wrong. Boolean variables are a bit similar with english questions. They always starts like this: "isLoading", "hasLoaded", "areItemsSame"

**Review by Assasinnys — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/basiccomponents/extensions/ViewExt.kt (hunk 10 above)
Comment (SergAndroid) [reply app/src/main/java/com/example/basiccomponents/extensions/ViewExt.kt]: In this case, the parameter is passed to the function "show" - true, hide - false. Show means the show command (показать). I don't know how to call it differently.

Comment (Assasinnys) [reply app/src/main/java/com/example/basiccomponents/extensions/ViewExt.kt]: It will be "isShowing"

Comment (Assasinnys) [reply app/src/main/java/com/example/basiccomponents/extensions/ViewExt.kt]: it's like a question. "Is showing? true/false". It helps you to read the code like a book :)

Comment (Assasinnys) [reply app/src/main/java/com/example/basiccomponents/extensions/ViewExt.kt]: Hm. I've found explanation for you. Look, you have name of a variable and value. So, your var's name is a question, value is answer to this question. )

Comment (SergAndroid) [reply app/src/main/java/com/example/basiccomponents/extensions/ViewExt.kt]: I agree that naming is very important, i am not a linguist, i may be wrong, but I think show is more suitable here: show? translates as "показать?"

**Review by Assasinnys — APPROVED** [review]

**Pull request merged**